cypress/integration/Password.feature

```diff
@@ -0,0 +1,51 @@
+Feature: Password 
+ 
+    I want to be able to reset or change my password 
+
+    # note that this requires Signup.feature to have run, to test resetting a password for an existing user
+    # todo The password email sent when a user doesn't exist yet, has a different title,
+    # isn't tested by Casper, and a test may need adding here
+    
+    @focus
+    # this scenario replicates 35-ChangePassword from Casper
+    # note that due to a bug in the system, this scenario can only be run before the forgot pasword scenario, not after
+    Scenario: Change existing password with invalid details
+        Given I ignore application exceptions
+        When I log in as standard test user
+        And I visit link containing "Your details"
+        Then I am taken to "/user/about-you"
+        When I visit link containing "Change Password"
+        Then I am taken to "/user/change-password"
+        When I try to change password to an invalid one
+        Then I am taken to "/user/change-password"
+        And I see "There was a problem changing your password" in the page text
+        And I see "Choose a new password that includes at least one digit (0-9)" in the page text
+        And I see "Choose a new password that includes at least one lower case letter (a-z)" in the page text
+        And I see "Choose a new password that includes at least one capital letter (A-Z)" in the page text
+        And I see "Enter matching passwords" in the page text
+  
+    @focus
+    Scenario: Reset Password using email link
+        Given I visit the login page
+        When I visit link containing "Forgotten your password?"
+        Then I am taken to "/forgot-password"
+        And I see "Reset your password" in the title
+        When I populate email fields with standard test user address
+        Then I see "We've emailed a link" in the page text
+        And I see standard test user in the page text
+        And I use password reset email to visit the link
+        When I choose a new password
+        Then I am taken to the login page
+        And I see "Password successfully reset" in the page text
+        When I log in with new password
+        # accept either type or dashboard page , so we're agnostic as to what has previously happened to this user
+        Then I am taken to the type or dashboard page
+        # change password back to old one. This wasn't in the original casper tests, but ensures this feature doesn't have any side effects 
+        When I visit link containing "Your details"
+        Then I am taken to "/user/about-you"
+        When I visit link containing "Change Password"
+        Then I am taken to "/user/change-password"
+        When I change password back to my old one
+        And I see "Your new password has been saved" in the page text
+
+
```

Comment:
These tests fail for me with:

```
Running:  Password.feature                                                                (5 of 5)


  Password
    1) Change existing password with invalid details
    2) Reset Password using email link


  0 passing (2m)
  2 failing

  1) Password
       Change existing password with invalid details:
     AssertionError: Timed out retrying: Expected to find content: 'Your details' but never did.
      at Context.eval (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:344:6)
      at Context.resolveAndRunStepDefinition (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:11940:9)
      at Context.eval (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:11281:35)

  2) Password
       Reset Password using email link:
     AssertionError: Timed out retrying: `cy.readFile("cypress/activation_emails/undefined.passwordreset")` failed because the file does not exist at the following path:

`/app/cypress/activation_emails/undefined.passwordreset`

https://on.cypress.io/readfile
      at openEmailAndVisitLink (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:81:6)
      at Context.eval (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:75:3)
      at Context.resolveAndRunStepDefinition (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:11940:9)
      at Context.eval (https://localhost:7002/__cypress/tests?p=cypress/integration/Password.feature:11281:35)
```

Any ideas?

Reply:
You have to run Password.feature after Signup.feature, during that run of the docker container. It resets the password for a user that has signed up. This is what the Casper tests did. Ultimately, as previously discussed as a group, we ought to have some code which, if it fails to find the signed up user, goes and makes that on the fly, although we ought to retain as part of the tests in CI, the scenario of a user being signed up via the website and everything being tested again after that.

Reply:
Thanks for clarifying. Does this mean we also need to switch these tests off if signup is off?

Reply:
In CI, we explicitly run the Signup tests (tests tagged Signup) followed by all tests not tagged Signup. When running locally through the GUI, you can easily do that by hand. I have now changed the cypress-local target in Makefile, so it does exactly what CI does, which sorts this.